### PR TITLE
box: store space name and engine name in `index_def` for error reporting

### PR DESCRIPTION
**Pull request by Astronomax**
This patch fixes a bug that was introduced by commit https://github.com/tarantool/tarantool/commit/8fc59b101c10cf71fbb157b6f98b64eb8155ae63 that adds helper functions for initializing payloads of DML errors related to key validation.

The problem was that after the space was deleted, a situation could occur that the `space.def->name` was still needed to generate an error message if the key format was incorrect in the `get()` call from `read_view`.

This is exactly what happened sometimes in the function `error_set_index`. One of the bug reproducers was to try to `get()` with an incorrect key format (`ER_EXACT_MATCH`) in `read_view` from deleted space.

The problem has been fixed: space name as well as engine name are also `index_def` fields now. There is no need to search for space and get these fields from it.

Also some other places where space name and engine name could be taken from `index_def` has been fixed.

Needed for tarantool/tarantool-ee#787

NO_DOC=bugfix
NO_TEST=ee
NO_CHANGELOG=bug not released

### CODE REVIEW

File: changelogs/unreleased/ghe-787-fix-null-pointer-dereference-in-exact_key_validate.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+## bugfix/box
+
+* Fixed a null pointer dereference when an `ER_EXACT_MATCH` error occurred when trying to `get()` in a `read_view` from deleted space. (ghe-787).
```
Comment (andreyaksenov) [inline changelogs/unreleased/ghe-787-fix-null-pointer-dereference-in-exact_key_validate.md:3]:
```suggestion
* Fixed a null pointer dereference when an `ER_EXACT_MATCH` error occurred on an attempt to call `get()` in a `read_view` from the deleted space. (ghe-787).
```

**Review by andreyaksenov — DISMISSED** [review]

File: src/box/space_def.h [hunk 2]
```diff
@@ -161,6 +161,8 @@ struct space_def {
 	uint32_t id;
 	/** User id of the creator of the space */
 	uint32_t uid;
+	/** Reference counter. */
+	int refs;
```
Comment (locker) [inline src/box/space_def.h:165]: Introducing reference counting for this structure is an overkill. Better just copy `space_name` to `index_def`.

**Review by locker — CHANGES_REQUESTED** [review]

File: src/box/index.cc [hunk 3]
```diff
@@ -99,8 +99,7 @@ error_set_index(struct error *error, const struct index_def *index_def)
 {
 	error_set_str(error, "index", index_def->name);
 	error_set_uint(error, "index_id", index_def->iid);
-	struct space *space = space_by_id(index_def->space_id);
-	error_set_str(error, "space", space->def->name);
+	error_set_str(error, "space", index_def->space_def->name);
```
Comment (locker) [inline src/box/index.cc:102]: There are a few other places where we look up a space just to get its name, for example:

https://github.com/tarantool/tarantool/blob/1f75231a0c0ce062e08660e0bc51485f7086412c/src/box/index.cc#L217-L221

Please find and fix them all.

**Comment by Astronomax** [comment]
Thanks for the review! Everything is fixed.
Perhaps it would be better to add space name to the signature of `index_def_new`?
```patch
struct index_def *
- index_def_new(uint32_t space_id, uint32_t iid, const char *name,
+ index_def_new(uint32_t space_id, char *space_name, uint32_t iid, const char *name,
	      uint32_t name_len, enum index_type type,
	      const struct index_opts *opts,
	      struct key_def *key_def, struct key_def *pk_def)
```
Looks like it's possible.

**Review by locker — CHANGES_REQUESTED** [review]

File: src/box/index.cc
[the remark this answers is not in the document]
Comment (Astronomax) [reply src/box/index.cc:1]: Fixed.

File: src/box/index_def.c [hunk 4]
```diff
@@ -144,6 +146,8 @@ index_def_dup(const struct index_def *def)
 	struct index_def *dup = xmalloc(sizeof(*dup));
 	*dup = *def;
 	dup->name = xstrdup(def->name);
+	if (def->space_name)
```
[the remark this answers is not in the document]
Comment (Astronomax) [reply src/box/index_def.c]: Fixed.

File: src/box/index_def.c [hunk 5]
```diff
@@ -35,6 +35,7 @@
 #include "json/json.h"
 #include "fiber.h"
 #include "tt_static.h"
+#include "space_cache.h"
```
[the remark this answers is not in the document]
Comment (Astronomax) [reply src/box/index_def.c]: Fixed.

File: changelogs/unreleased/ghe-787-fix-null-pointer-dereference-in-exact_key_validate.md [hunk 6]
```diff
@@ -0,0 +1,3 @@
+## bugfix/box
+
+* Fixed a null pointer dereference when an `ER_EXACT_MATCH` error occurred on an attempt to call `get()` in a `read_view` from the deleted space. (ghe-787).
```
[the remark this answers is not in the document]
Comment (Astronomax) [reply changelogs/unreleased/ghe-787-fix-null-pointer-dereference-in-exact_key_validate.md]: Fixed.

File: src/box/index_def.h
Comment (locker) [inline src/box/index_def.h:1]: The commit message doesn't really shed the light on what was wrong and how it was fixed:

> ```
> box: fix null pointer dereference in exact_key_validate
> 
> Fixed a null pointer dereference in `exact_key_validate`. When you tried
> to `get()` with an incorrect key format (`ER_EXACT_MATCH`) in `read_view`
> from deleted space, a null pointer was dereferenced and tarantool crashed.
> ```

Actually, it's quite confusing, because the commit doesn't actually fix a null pointer dereference in *CE*.

Please rewrite both the subject line and the message so that it's clear what the commit does, why the issue appeared (reference the bad commit), and how this commit helps fixing it:

```
box: store space name and engine name in index_def for error reporting

<what's the problem, why it appeared, how this commit fixes it>
```

File: src/box/vinyl.c [hunk 7]
```diff
@@ -1627,12 +1623,13 @@ vy_check_is_unique_secondary(struct vy_tx *tx, const struct vy_read_view **rv,
 {
 	assert(lsm->opts.is_unique);
 	if (!lsm->cmp_def->is_multikey) {
-		return vy_check_is_unique_secondary_one(tx, rv, lsm, stmt,
-							MULTIKEY_NONE);
+		return vy_check_is_unique_secondary_one(tx, rv,
+				lsm, stmt, MULTIKEY_NONE);
```
Comment (locker) [inline src/box/vinyl.c:1627]: Nit: Bad indentation - the continuation should be aligned by the parenthesis. I fixed checkpatch - now it detects this error.

File: src/box/vinyl.c [hunk 8]
```diff
@@ -1627,12 +1623,13 @@ vy_check_is_unique_secondary(struct vy_tx *tx, const struct vy_read_view **rv,
...
 	}
 	int count = tuple_multikey_count(stmt, lsm->cmp_def);
 	for (int i = 0; i < count; ++i) {
-		if (vy_check_is_unique_secondary_one(tx, rv, lsm, stmt, i) != 0)
+		if (vy_check_is_unique_secondary_one(tx, rv, lsm, stmt, i)
+			!= 0)
```
Comment (locker) [inline src/box/vinyl.c:1632]: Nit: Unrelated change. Please revert.

Comment (Astronomax) [reply src/box/vinyl.c]: Fixed.

File: src/box/vinyl.c (hunk 7 above)
Comment (Astronomax) [reply src/box/vinyl.c]: Fixed.

File: src/box/index_def.h
Comment (Astronomax) [reply src/box/index_def.h:1]: Fixed.

**Review by locker — APPROVED** [review]

**Comment by locker** [comment]
@Astronomax The `debug_asan_clang` workflow found some leaks. PTAL.

```
[011] master | Direct leak of 8 byte(s) in 1 object(s) allocated from:
[011] master |     #0 0x55cc2e5f2b8d in strdup (/__w/tarantool/tarantool/src/tarantool+0x11b8b8d) (BuildId: f223e8a7b8667551d8b411188d51355862c71e25)
[011] master |     #1 0x55cc2e728709 in index_def_dup /__w/tarantool/tarantool/src/box/index_def.c:150:21
[011] master |     #2 0x55cc2e719713 in index_create /__w/tarantool/tarantool/src/box/index.cc:780:8
[011] master |     #3 0x55cc2e753896 in index* memtx_tree_index_new_tpl<true>(memtx_engine*, index_def*, index_vtab const*) /__w/tarantool/tarantool/src/box/memtx_tree.cc:2283:2
[011] master |     #4 0x55cc2e753623 in memtx_tree_index_new /__w/tarantool/tarantool/src/box/memtx_tree.cc:2323:10
[011] master |     #5 0x55cc2e970eee in memtx_space_create_index /__w/tarantool/tarantool/src/box/memtx_space.c:933:10
[011] master |     #6 0x55cc2ed76cfa in space_create_index /__w/tarantool/tarantool/src/box/space.h:605:9
[011] master |     #7 0x55cc2ed73c32 in space_create /__w/tarantool/tarantool/src/box/space.c:584:25
[011] master |     #8 0x55cc2e9619ac in memtx_space_new /__w/tarantool/tarantool/src/box/memtx_space.c:1502:6
[011] master |     #9 0x55cc2e92ee44 in memtx_engine_create_space(engine*, space_def*, rlist*) /__w/tarantool/tarantool/src/box/memtx_engine.cc:564:9
[011] master |     #10 0x55cc2ed7b1ee in engine_create_space /__w/tarantool/tarantool/src/box/engine.h:365:9
[011] master |     #11 0x55cc2ed7ad39 in space_new /__w/tarantool/tarantool/src/box/space.c:686:9
[011] master |     #12 0x55cc2ee85afc in space_new_xc(space_def*, rlist*) /__w/tarantool/tarantool/src/box/space.h:772:24
[011] master |     #13 0x55cc2ee7c550 in alter_space_do(txn_stmt*, alter_space*) /__w/tarantool/tarantool/src/box/alter.cc:1013:21
[011] master |     #14 0x55cc2ee48a39 in on_replace_dd_space(trigger*, void*) /__w/tarantool/tarantool/src/box/alter.cc:2395:4
[011] master |     #15 0x55cc2fa82236 in trigger_run_list(rlist*, void*) /__w/tarantool/tarantool/src/lib/core/trigger.cc:100:7
[011] master |     #16 0x55cc2fa807da in trigger_run /__w/tarantool/tarantool/src/lib/core/trigger.cc:133:9
[011] master |     #17 0x55cc2ed80924 in space_on_replace /__w/tarantool/tarantool/src/box/space.c:1001:11
[011] master |     #18 0x55cc2ef1632c in txn_commit_stmt /__w/tarantool/tarantool/src/box/txn.c:678:13
[011] master |     #19 0x55cc2ef7f995 in box_process_rw /__w/tarantool/tarantool/src/box/box.cc:518:6
[011] master |     #20 0x55cc2eff1ce0 in wal_stream_apply_dml_row(wal_stream*, xrow_header*) /__w/tarantool/tarantool/src/box/box.cc:830:7
[011] master |     #21 0x55cc2efec692 in wal_stream_apply_row(xstream*, xrow_header*) /__w/tarantool/tarantool/src/box/box.cc:1048:13
[011] master |     #22 0x55cc2f0ed471 in xstream_write /__w/tarantool/tarantool/src/box/xstream.cc:39:3
[011] master |     #23 0x55cc2f0e4cf0 in recover_xlog(recovery*, xstream*, vclock const*) /__w/tarantool/tarantool/src/box/recovery.cc:305:7
[011] master |     #24 0x55cc2f0e14e2 in recover_remaining_wals(recovery*, xstream*, vclock const*, bool) /__w/tarantool/tarantool/src/box/recovery.cc:373:3
[011] master |     #25 0x55cc2efe41a3 in local_recovery(vclock const*) /__w/tarantool/tarantool/src/box/box.cc:5401:2
[011] master |     #26 0x55cc2efc5b29 in box_cfg_xc() /__w/tarantool/tarantool/src/box/box.cc:5617:3
[011] master |     #27 0x55cc2efc4c2f in box_cfg /__w/tarantool/tarantool/src/box/box.cc:5754:3
[011] master |     #28 0x55cc2e648d05 in load_cfg /__w/tarantool/tarantool/src/main.cc:507:2
[011] master |     #29 0x55cc2f6ef188 in lbox_cfg_load(lua_State*) /__w/tarantool/tarantool/src/box/lua/cfg.cc:61:3
```

**Comment by Astronomax** [comment]
Fixed.
```diff
diff --git a/src/box/space.c b/src/box/space.c
index 5e6362488..b17427b70 100644
--- a/src/box/space.c
+++ b/src/box/space.c
@@ -578,10 +578,11 @@ space_create(struct space *space, struct engine *engine,
        }
 
        rlist_foreach_entry(index_def, key_list, link) {
-               index_def->space_name = xstrdup(def->name);
-               strlcpy(index_def->engine_name, def->engine_name,
-                       ENGINE_NAME_MAX);
                struct index *index = space_create_index(space, index_def);
+               free(index->def->space_name);
+               index->def->space_name = xstrdup(def->name);
+               strlcpy(index->def->engine_name, def->engine_name,
+                       ENGINE_NAME_MAX);
                if (index == NULL)
                        goto fail_free_indexes;
                space->index_map[index_def->iid] = index;
```

**Comment by Astronomax** [comment]
Fixed.
```diff
diff --git a/home/astronomax/space.c b/src/box/space.c
index 2322e5228..bd1c04b6b 100644
--- a/home/astronomax/space.c
+++ b/src/box/space.c
@@ -578,10 +578,13 @@ space_create(struct space *space, struct engine *engine,
        }
 
        rlist_foreach_entry(index_def, key_list, link) {
-               index_def->space_name = xstrdup(def->name);
-               strlcpy(index_def->engine_name, def->engine_name,
+               struct index_def *temp_def = index_def_dup(index_def);
+               free(temp_def->space_name);
+               temp_def->space_name = xstrdup(def->name);
+               strlcpy(temp_def->engine_name, def->engine_name,
                        ENGINE_NAME_MAX);               
                struct index *index = space_create_index(space, temp_def);
+               index_def_delete(temp_def);
                if (index == NULL)
                        goto fail_free_indexes;
                space->index_map[index_def->iid] = index;
```

File: src/box/space.c [hunk 9]
```diff
@@ -576,8 +576,15 @@ space_create(struct space *space, struct engine *engine,
...
+
 	rlist_foreach_entry(index_def, key_list, link) {
-		struct index *index = space_create_index(space, index_def);
+		struct index_def *temp_def = index_def_dup(index_def);
+		free(temp_def->space_name);
+		temp_def->space_name = xstrdup(def->name);
+		strlcpy(temp_def->engine_name, def->engine_name,
+			ENGINE_NAME_MAX);		
+		struct index *index = space_create_index(space, temp_def);
+		index_def_delete(temp_def);
```
Comment (locker) [inline src/box/space.c:587]: This looks rather ugly :-/

Maybe, better add the two new arguments `space_name` and `engine_name` to `index_def_new()`? There aren't that many places where it's called so patching them all shouldn't be painful. The benefit of it is that we wouldn't have to think about possible `index_def::space_name` and `index_def::engine_name` changes - they'd stay the same the whole time until the object is deleted.

A side note: There's the `asan-ci` label you could use instead of `full-ci`. In contrast to `full-ci`, `asan-ci` only triggers ASAN workflows so CI should pass faster.

File: src/box/index_def.c [hunk 10]
```diff
@@ -94,6 +94,7 @@ const struct opt_def index_opts_reg[] = {
 
 struct index_def *
 index_def_new(uint32_t space_id, uint32_t iid, const char *name,
+	      const char *space_name, const char* engine_name,
 	      uint32_t name_len, enum index_type type,
 	      const struct index_opts *opts,
 	      struct key_def *key_def, struct key_def *pk_def)
```
Comment (locker) [inline src/box/index_def.c:100]: Please keep `name` and `name_len` together in the argument list:

```C
struct index_def *
index_def_new(uint32_t space_id, uint32_t iid,
	      const char *name, uint32_t name_len,
	      const char *space_name, const char *engine_name,
	      enum index_type type, const struct index_opts *opts,
	      struct key_def *key_def, struct key_def *pk_def)
```

File: src/box/index_def.c [hunk 11]
```diff
@@ -107,14 +108,23 @@ index_def_new(uint32_t space_id, uint32_t iid, const char *name,
 	}
 	def->name = strndup(name, name_len);
 	if (def->name == NULL) {
-		index_def_delete(def);
 		diag_set(OutOfMemory, name_len + 1, "malloc", "index_def name");
-		return NULL;
+		goto error;
```
Comment (locker) [inline src/box/index_def.c:112]: While you are at it, please make `index_def_new` use `xstrdup` and `xmalloc` and remove all checks if it returns `NULL`.

File: src/box/space.c [hunk 12]
```diff
@@ -576,8 +576,16 @@ space_create(struct space *space, struct engine *engine,
...
 	rlist_foreach_entry(index_def, key_list, link) {
-		struct index *index = space_create_index(space, index_def);
+		struct index_def *temp_def =
+			index_def_new(index_def->space_id, index_def->iid,
+				      index_def->name, def->name,
+				      def->engine_name, strlen(index_def->name),
+				      index_def->type, &index_def->opts,
+				      index_def->key_def, index_def->pk_def);
+		struct index *index = space_create_index(space, temp_def);
+		index_def_delete(temp_def);
```
Comment (locker) [inline src/box/space.c:588]: You should create `index_def` with a proper space and engine names in `index_def_new_from_tuple` instead of duplicating it here.

**Review by locker — CHANGES_REQUESTED** [review]

File: src/box/space.c (hunk 12 above)
Comment (Astronomax) [reply src/box/space.c]: Fixed.

File: src/box/index_def.c (hunk 11 above)
Comment (Astronomax) [reply src/box/index_def.c]: Fixed.

File: src/box/index_def.c (hunk 10 above)
Comment (Astronomax) [reply src/box/index_def.c]: Fixed.

**Comment by Astronomax** [comment]
```diff
diff --git a/perf/memtx.cc b/perf/memtx.cc
index bb13cc27e..b28d87ed6 100644
--- a/perf/memtx.cc
+++ b/perf/memtx.cc
@@ -160,9 +160,10 @@ class Memtx final {
 		if (key_def == nullptr)
 			panic("failed to create new key definition");
 		tree_idx_def = ::index_def_new(sid, tree_index_id, "pk",
-					       NULL, NULL,
-					       std::strlen("pk"), TREE,
-					       &idx_opts, key_def,
+					       std::strlen("pk"),
+					       space_def->name,
+					       space_def->engine_name,
+					       TREE, &idx_opts, key_def,
 					       /*pk_def=*/nullptr);
 		if (tree_idx_def == nullptr)
 			panic("failed to create new index definition");
diff --git a/src/box/alter.cc b/src/box/alter.cc
index f2649ee37..48b639c02 100644
--- a/src/box/alter.cc
+++ b/src/box/alter.cc
@@ -389,8 +389,9 @@ index_def_new_from_tuple(struct tuple *tuple, struct space *space)
 	if (key_def == NULL)
 		return NULL;
 	struct index_def *index_def =
-		index_def_new(id, index_id, name, NULL, NULL, name_len, type,
-			      &opts, key_def, space_index_key_def(space, 0));
+		index_def_new(id, index_id, name, name_len, space->def->name,
+			      space->def->engine_name, type, &opts, key_def,
+			      space_index_key_def(space, 0));
 	if (index_def == NULL)
 		return NULL;
 	auto index_def_guard = make_scoped_guard([=] { index_def_delete(index_def); });
@@ -1767,9 +1768,10 @@ alter_space_move_indexes(struct alter_space *alter, uint32_t begin,
 		 * Rebuild secondary indexes that depend on the
 		 * primary key since primary key parts have changed.
 		 */
-		new_def = index_def_new(old_def->space_id, old_def->iid, old_def->name,
-					old_def->space_name, old_def->engine_name,
-					strlen(old_def->name),
+		new_def = index_def_new(old_def->space_id, old_def->iid,
+					old_def->name, strlen(old_def->name),
+					old_def->space_name,
+					old_def->engine_name,
 					old_def->type, &old_def->opts,
 					old_def->key_def, alter->pk_def);
 		index_def_update_optionality(new_def, min_field_count);
diff --git a/src/box/index_def.c b/src/box/index_def.c
index 43418d5ed..e30ff8aac 100644
--- a/src/box/index_def.c
+++ b/src/box/index_def.c
@@ -94,37 +94,24 @@ const struct opt_def index_opts_reg[] = {
 
 struct index_def *
 index_def_new(uint32_t space_id, uint32_t iid, const char *name,
-	      const char *space_name, const char* engine_name,
-	      uint32_t name_len, enum index_type type,
+	      uint32_t name_len, const char *space_name,
+	      const char *engine_name, enum index_type type,
 	      const struct index_opts *opts,
 	      struct key_def *key_def, struct key_def *pk_def)
 {
 	assert(name_len <= BOX_NAME_MAX);
 	/* Use calloc to make index_def_delete() safe at all times. */
-	struct index_def *def = (struct index_def *) calloc(1, sizeof(*def));
-	if (def == NULL) {
-		diag_set(OutOfMemory, sizeof(*def), "malloc", "struct index_def");
+	struct index_def *def = xcalloc(1, sizeof(*def));
+	def->name = xstrndup(name, name_len);
+	if (identifier_check(def->name, name_len)) {
+		index_def_delete(def);
 		return NULL;
 	}
-	def->name = strndup(name, name_len);
-	if (def->name == NULL) {
-		diag_set(OutOfMemory, name_len + 1, "malloc", "index_def name");
-		goto error;
-	}
-	if (identifier_check(def->name, name_len))
-		goto error;
-	if (space_name != NULL) {
-		def->space_name = strdup(space_name);
-		if (def->space_name == NULL) {
-			diag_set(OutOfMemory, strlen(space_name) + 1, "malloc",
-				 "index_def space_name");
-			goto error;
-		}
-	}
+	if (space_name != NULL)
+		def->space_name = xstrdup(space_name);
 	memset(def->engine_name, 0, sizeof(def->engine_name));
 	if (engine_name != NULL)
 		strlcpy(def->engine_name, engine_name, ENGINE_NAME_MAX);
-
 	def->key_def = key_def_dup(key_def);
 	if (iid != 0) {
 		assert(pk_def != NULL);
@@ -146,10 +133,6 @@ index_def_new(uint32_t space_id, uint32_t iid, const char *name,
 	def->iid = iid;
 	def->opts = *opts;
 	return def;
-
-error:
-	index_def_delete(def);
-	return NULL;
 }
 
 struct index_def *
diff --git a/src/box/index_def.h b/src/box/index_def.h
index 8c847e523..871c1b269 100644
--- a/src/box/index_def.h
+++ b/src/box/index_def.h
@@ -280,8 +280,8 @@ index_def_list_add(struct rlist *index_def_list, struct index_def *index_def)
  */
 struct index_def *
 index_def_new(uint32_t space_id, uint32_t iid, const char *name,
-	      const char *space_name, const char* engine_name,
-	      uint32_t name_len, enum index_type type,
+	      uint32_t name_len, const char *space_name,
+	      const char *engine_name, enum index_type type,
 	      const struct index_opts *opts,
 	      struct key_def *key_def, struct key_def *pk_def);
 
diff --git a/src/box/schema.cc b/src/box/schema.cc
index d385936c1..7ec4f3ee9 100644
--- a/src/box/schema.cc
+++ b/src/box/schema.cc
@@ -211,6 +211,11 @@ sc_space_new(uint32_t id, const char *name,
 	     uint32_t key_part_count,
 	     struct trigger *replace_trigger)
 {
+	struct space_def *def =
+		space_def_new_xc(id, ADMIN, 0, name, strlen(name), "memtx",
+				 strlen("memtx"), &space_opts_default, NULL, 0,
+				 NULL, 0);
+	auto def_guard = make_scoped_guard([=] { space_def_delete(def); });
 	struct key_def *key_def = key_def_new(key_parts, key_part_count, 0);
 	if (key_def == NULL)
 		diag_raise();
@@ -219,9 +224,9 @@ sc_space_new(uint32_t id, const char *name,
 	struct index_def *index_def = index_def_new(id, /* space id */
 						    0 /* index id */,
 						    "primary", /* name */
-						    NULL, /* space name */
-						    NULL, /* engine name */
 						    strlen("primary"),
+						    def->name, /* space name */
+						    def->engine_name,
 						    TREE /* index type */,
 						    &index_opts_default,
 						    key_def, NULL);
@@ -229,11 +234,6 @@ sc_space_new(uint32_t id, const char *name,
 		diag_raise();
 	auto index_def_guard =
 		make_scoped_guard([=] { index_def_delete(index_def); });
-	struct space_def *def =
-		space_def_new_xc(id, ADMIN, 0, name, strlen(name), "memtx",
-				 strlen("memtx"), &space_opts_default, NULL, 0,
-				 NULL, 0);
-	auto def_guard = make_scoped_guard([=] { space_def_delete(def); });
 	struct rlist key_list;
 	rlist_create(&key_list);
 	rlist_add_entry(&key_list, index_def, link);
diff --git a/src/box/space.c b/src/box/space.c
index 324e86b87..1d2876ff8 100644
--- a/src/box/space.c
+++ b/src/box/space.c
@@ -578,14 +578,7 @@ space_create(struct space *space, struct engine *engine,
 	}
 
 	rlist_foreach_entry(index_def, key_list, link) {
-		struct index_def *temp_def =
-			index_def_new(index_def->space_id, index_def->iid,
-				      index_def->name, def->name,
-				      def->engine_name, strlen(index_def->name),
-				      index_def->type, &index_def->opts,
-				      index_def->key_def, index_def->pk_def);
-		struct index *index = space_create_index(space, temp_def);
-		index_def_delete(temp_def);
+		struct index *index = space_create_index(space, index_def);
 		if (index == NULL)
 			goto fail_free_indexes;
 		space->index_map[index_def->iid] = index;
diff --git a/src/box/sql.c b/src/box/sql.c
index e55ffad98..ba274f739 100644
--- a/src/box/sql.c
+++ b/src/box/sql.c
@@ -378,29 +378,34 @@ sql_ephemeral_space_new(const struct sql_space_info *info)
 		parts[i].coll_id = info->coll_ids[j];
 	}
 
+	struct space_def *space_def = space_def_new_ephemeral(field_count,
+							      fields);
+
+	if (space_def == NULL)
+		return NULL;
+
 	struct key_def *key_def = key_def_new(parts, part_count, 0);
-	if (key_def == NULL)
+	if (key_def == NULL) {
+		space_def_delete(space_def);
 		return NULL;
+	}
 
 	const char *name = "ephemer_idx";
-	struct index_def *index_def = index_def_new(0, 0, name, NULL, NULL,
-						    strlen(name), TREE, &index_opts_default,
+	struct index_def *index_def = index_def_new(0, 0, name, strlen(name),
+						    space_def->name,
+						    space_def->engine_name,
+						    TREE, &index_opts_default,
 						    key_def, NULL);
 	key_def_delete(key_def);
-	if (index_def == NULL)
+	if (index_def == NULL) {
+		space_def_delete(space_def);
 		return NULL;
+	}
 
 	struct rlist key_list;
 	rlist_create(&key_list);
 	rlist_add_entry(&key_list, index_def, link);
 
-	struct space_def *space_def = space_def_new_ephemeral(field_count,
-							      fields);
-	if (space_def == NULL) {
-		index_def_delete(index_def);
-		return NULL;
-	}
-
 	struct space *space = space_new_ephemeral(space_def, &key_list);
 	index_def_delete(index_def);
 	space_def_delete(space_def);
diff --git a/src/box/sql/build.c b/src/box/sql/build.c
index 6e58151d3..ef0cb1967 100644
--- a/src/box/sql/build.c
+++ b/src/box/sql/build.c
@@ -2654,9 +2654,9 @@ index_fill_def(struct Parse *parse, struct index *index,
 	 * only for comparison routine. Meanwhile on front-end
 	 * side only definition is used.
 	 */
-	index->def = index_def_new(space_def->id, 0, name,
+	index->def = index_def_new(space_def->id, 0, name, name_len,
 				   space_def->name, space_def->engine_name,
-				   name_len, TREE, &opts, key_def, NULL);
+				   TREE, &opts, key_def, NULL);
 	if (index->def == NULL)
 		goto tnt_error;
 	index->def->iid = iid;
diff --git a/src/box/sql/where.c b/src/box/sql/where.c
index 725399cb5..161f7b82e 100644
--- a/src/box/sql/where.c
+++ b/src/box/sql/where.c
@@ -895,10 +895,10 @@ constructAutomaticIndex(Parse * pParse,			/* The parsing context */
 	index_opts_create(&opts);
 	const char *idx_name = "ephemeral index";
 	struct index_def *idx_def = index_def_new(space->def->id, 0, idx_name,
+						  strlen(idx_name),
 						  space->def->name,
 						  space->def->engine_name,
-						  strlen(idx_name), TREE, &opts,
-						  key_def, NULL);
+						  TREE, &opts, key_def, NULL);
 	key_def_delete(key_def);
 	if (idx_def == NULL) {
 		pParse->is_aborted = true;
@@ -2136,9 +2136,9 @@ whereLoopAddBtree(WhereLoopBuilder * pBuilder,	/* WHERE clause information */
 		struct index_opts opts;
 		index_opts_create(&opts);
 		fake_index = index_def_new(space->def->id, 0, "fake_autoindex",
+					   sizeof("fake_autoindex") - 1,
 					   space->def->name,
 					   space->def->engine_name,
-					   sizeof("fake_autoindex") - 1,
 					   TREE, &opts, key_def, NULL);
 		key_def_delete(key_def);
 		if (fake_index == NULL)
diff --git a/test/luatest.log b/test/luatest.log
new file mode 100644
index 000000000..e69de29bb
diff --git a/test/unit/vy_point_lookup.c b/test/unit/vy_point_lookup.c
index b43ae6ed1..0415d6ed4 100644
--- a/test/unit/vy_point_lookup.c
+++ b/test/unit/vy_point_lookup.c
@@ -94,8 +94,8 @@ test_basic()
 
 	struct index_opts index_opts = index_opts_default;
 	struct index_def *index_def =
-		index_def_new(512, 0, "primary", NULL, NULL,
-			      sizeof("primary") - 1, TREE,
+		index_def_new(512, 0, "primary", sizeof("primary") - 1,
+			      NULL, NULL, TREE,
 			      &index_opts, key_def, NULL);
 
 	struct vy_lsm *pk = vy_lsm_new(&lsm_env, &cache_env, &mem_env,
```

**Review by locker — APPROVED** [review]
